File: apps/clinician_flutter/lib/domain/gaps.dart
/// HANDOFF §2.8 — Gap detection is rules, not a model.
///
/// "Is the medication field empty?" is a schema check against the configured
/// template: deterministic, unit-testable, reproducible, auditable.
///
/// HANDOFF §4.3 — output wording is neutral. "Not mentioned: X", never
/// "Critical information missing": ranking which omission matters is a
/// clinical judgement, and this system never makes one.
/// Dart mirror of packages/rules/src/gaps.ts.
library;

class TemplateField {
  final String id;
  final String label;
  final bool required;

  /// Routing keywords used by the (swappable) structurer. Config data, not logic.
  final List<String> keywords;

  const TemplateField({
    required this.id,
    required this.label,
    required this.required,
    required this.keywords,
  });

  factory TemplateField.fromMap(Map<String, dynamic> m) => TemplateField(
        id: m['id'] as String,
        label: m['label'] as String,
        required: m['required'] as bool? ?? false,
        keywords: (m['keywords'] as List<dynamic>? ?? const []).cast<String>(),
      );

  Map<String, dynamic> toMap() =>
      {'id': id, 'label': label, 'required': required, 'keywords': keywords};
}

class FieldContent {
  final String fieldId;
  final String text;

  const FieldContent({required this.fieldId, required this.text});

  factory FieldContent.fromMap(Map<String, dynamic> m) => FieldContent(
      fieldId: m['fieldId'] as String, text: m['text'] as String? ?? '');

  Map<String, dynamic> toMap() => {'fieldId': fieldId, 'text': text};
}

class OmissionFlagData {
  final String fieldId;
  final String label;

  /// Always the neutral form: `Not mentioned: $label`.
  final String message;

  const OmissionFlagData(
      {required this.fieldId, required this.label, required this.message});
}

List<OmissionFlagData> detectGaps(
    List<TemplateField> template, List<FieldContent> contents) {
  final filled = {for (final c in contents) c.fieldId: c.text};

  final flags = <OmissionFlagData>[];
  // Template order, not importance order — importance is a clinical judgement.
  for (final field in template) {
    if (!field.required) continue;
    final text = filled[field.id];
    if (text == null || text.trim().isEmpty) {
      flags.add(OmissionFlagData(
        fieldId: field.id,
        label: field.label,
        message: 'Not mentioned: ${field.label}',
      ));
    }
  }
  return flags;
}

/// Share of template fields with any content, 0..1.
double fileCompleteness(
    List<TemplateField> template, List<FieldContent> contents) {
  if (template.isEmpty) return 1;
  final filled = contents
      .where((c) => c.text.trim().isNotEmpty)
      .map((c) => c.fieldId)
      .toSet();
  final covered = template.where((f) => filled.contains(f.id)).length;
  return covered / template.length;
}

File: packages/rules/src/gaps.ts
/**
 * HANDOFF §2.8 — Gap detection is rules, not a model.
 *
 * "Is the medication field empty?" is a schema check against the configured
 * template: deterministic, unit-testable, reproducible, auditable.
 *
 * HANDOFF §4.3 — output wording is neutral. "Not mentioned: X", never
 * "Critical information missing": ranking which omission matters is a
 * clinical judgement, and this system never makes one.
 */

export interface TemplateField {
  id: string
  label: string
  required: boolean
  /** Routing keywords used by the (swappable) structurer. Config data, not logic. */
  keywords: string[]
}

export interface FieldContent {
  fieldId: string
  text: string
}

export interface OmissionFlag {
  fieldId: string
  label: string
  /** Always the neutral form: `Not mentioned: ${label}`. */
  message: string
}

export function detectGaps(
  template: readonly TemplateField[],
  contents: readonly FieldContent[],
): OmissionFlag[] {
  const filled = new Map<string, string>()
  for (const c of contents) filled.set(c.fieldId, c.text)

  const flags: OmissionFlag[] = []
  // Template order, not importance order — importance is a clinical judgement.
  for (const field of template) {
    if (!field.required) continue
    const text = filled.get(field.id)
    if (text === undefined || text.trim() === '') {
      flags.push({
        fieldId: field.id,
        label: field.label,
        message: `Not mentioned: ${field.label}`,
      })
    }
  }
  return flags
}

/** Share of template fields with any content, 0..1 — the queue's second sort key. */
export function fileCompleteness(
  template: readonly TemplateField[],
  contents: readonly FieldContent[],
): number {
  if (template.length === 0) return 1
  const filled = new Set(
    contents.filter((c) => c.text.trim() !== '').map((c) => c.fieldId),
  )
  const covered = template.filter((f) => filled.has(f.id)).length
  return covered / template.length
}
